src/common/threadpool.cpp

```diff
@@ -60,11 +62,13 @@ threadpool::~threadpool() {
 void threadpool::submit(waiter *obj, std::function<void()> f) {
   entry e = {obj, f};
   boost::unique_lock<boost::mutex> lock(mutex);
-  if (active == max && !queue.empty()) {
+  if ((active == max && !queue.empty()) || depth > 0)
     // if all available threads are already running
     // and there's work waiting, just run in current thread
     lock.unlock();
+    ++depth;
```

Comment:
Might as well move the increment/decrement of depth outside the if since it's occurring in both halves. And then change "depth > 0" to "depth > 1".